File: packages/oxlint-plugin-react-doctor/src/plugin/rules/r3f/three-no-ignored-basic-material-properties.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getStaticObjectPropertyValue } from "../../utils/get-static-object-property-value.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { MESH_BASIC_MATERIAL_IGNORED_PBR_PROPERTY_NAMES } from "./constants.js";
import { getApiReferenceProvenance } from "./utils/get-api-reference-provenance.js";
import { isThreeModuleSource } from "./utils/is-three-module-source.js";
import { getThreePropertyAssignment } from "./utils/get-three-property-assignment.js";

export const threeNoIgnoredBasicMaterialProperties = defineRule({
  id: "three-no-ignored-basic-material-properties",
  title: "PBR property ignored by Three.js MeshBasicMaterial",
  category: "Correctness",
  severity: "warn",
  recommendation:
    "Use MeshStandardMaterial or MeshPhysicalMaterial when a mesh needs roughness or metalness",
  create: (context: RuleContext) => ({
    NewExpression(node: EsTreeNodeOfType<"NewExpression">) {
      const provenance = getApiReferenceProvenance(node.callee, context.scopes);
      const parameters = node.arguments[0];
      if (
        provenance?.apiName !== "MeshBasicMaterial" ||
        !isThreeModuleSource(provenance.moduleSource) ||
        !parameters
      ) {
        return;
      }
      for (const propertyName of MESH_BASIC_MATERIAL_IGNORED_PBR_PROPERTY_NAMES) {
        const propertyValue = getStaticObjectPropertyValue(parameters, propertyName);
        if (!propertyValue) continue;
        context.report({
          node: propertyValue,
          message: `MeshBasicMaterial ignores ${propertyName} because it is not a PBR material. Use MeshStandardMaterial or MeshPhysicalMaterial for this property`,
        });
      }
    },
    AssignmentExpression(node: EsTreeNodeOfType<"AssignmentExpression">) {
      const assignment = getThreePropertyAssignment(node, context);
      if (
        assignment?.constructorName !== "MeshBasicMaterial" ||
        !MESH_BASIC_MATERIAL_IGNORED_PBR_PROPERTY_NAMES.has(assignment.propertyName)
      ) {
        return;
      }
      context.report({
        node: assignment.value,
        message: `MeshBasicMaterial ignores ${assignment.propertyName} because it is not a PBR material. Use MeshStandardMaterial or MeshPhysicalMaterial for this property`,
      });
    },
  }),
});

File: native/oxlint/rules/three-no-ignored-basic-material-properties.rs
use oxc_ast::AstKind;
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_span::GetSpan;
use oxc_syntax::operator::AssignmentOperator;

use crate::{AstNode, context::LintContext, rule::Rule};

const IGNORED_PROPERTY_NAMES: [&str; 2] = ["metalness", "roughness"];
const MESH_BASIC_MATERIAL_CONSTRUCTOR_NAMES: [&str; 1] = ["MeshBasicMaterial"];

#[derive(Debug, Default, Clone)]
pub struct ThreeNoIgnoredBasicMaterialProperties;

declare_oxc_lint!(
    /// Disallow PBR-only properties on Three.js basic materials.
    ThreeNoIgnoredBasicMaterialProperties,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Disallow ignored PBR properties on basic materials.",
);

impl Rule for ThreeNoIgnoredBasicMaterialProperties {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        match node.kind() {
            AstKind::NewExpression(new_expression)
                if three_module_api_path_matches(
                    &new_expression.callee,
                    &MESH_BASIC_MATERIAL_CONSTRUCTOR_NAMES,
                    ctx,
                ) =>
            {
                let Some(parameters) = new_expression
                    .arguments
                    .first()
                    .and_then(oxc_ast::ast::Argument::as_expression)
                else {
                    return;
                };
                for property_name in IGNORED_PROPERTY_NAMES {
                    let Some(property_value) =
                        get_static_object_property_value(parameters, property_name)
                    else {
                        continue;
                    };
                    report_ignored_basic_material_property(property_name, property_value, ctx);
                }
            }
            AstKind::AssignmentExpression(assignment)
                if assignment.operator == AssignmentOperator::Assign =>
            {
                let Some(member_expression) = assignment.left.as_member_expression() else {
                    return;
                };
                let Some(property_name) = member_expression.static_property_name() else {
                    return;
                };
                if !IGNORED_PROPERTY_NAMES.contains(&property_name)
                    || three_constructor_name(
                        member_expression.object(),
                        &MESH_BASIC_MATERIAL_CONSTRUCTOR_NAMES,
                        ctx,
                    )
                    .is_none()
                {
                    return;
                }
                report_ignored_basic_material_property(property_name, &assignment.right, ctx);
            }
            _ => {}
        }
    }
}

fn report_ignored_basic_material_property(
    property_name: &str,
    property_value: &oxc_ast::ast::Expression,
    ctx: &LintContext,
) {
    ctx.diagnostic(
        OxcDiagnostic::warn(format!(
            "MeshBasicMaterial ignores {property_name} because it is not a PBR material. Use MeshStandardMaterial or MeshPhysicalMaterial for this property"
        ))
        .with_label(property_value.span()),
    );
}
